File: autoschematic-sdk-ts/src/addr.ts
/**
 * Turn a template containing [groupName] placeholders into a RegExp
 * whose groups can be read via `match(...).groups?.groupName`.
 *
 * Special regex metacharacters in the static parts are escaped automatically.
 * Each placeholder becomes   (?<groupName>[^/]+)
 *
 * Example:
 *   const rx = matchAddr("snowflake/warehouses/$[name].sql");
 *   const m  = "snowflake/warehouses/WH1.sql".match(rx);
 *   console.error(m?.groups?.name);   // -> "WH1"
 */
export const matchAddr = (template: string): RegExp => {
  // Escape any chars that may be interpreted as a regex
  const escape = (s: string) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

  // Start of regex
  let rx = '^';

  for (let i = 0; i < template.length; i++) {
    // Look for "["...
    if (template[i] === '[') {
      i += 1;
      const start = i;
      // ...look for "]"
      while (i < template.length && template[i] !== ']') i++;
      // Check if we hit the end before we found "]"
      if (i === template.length) throw new Error('matchAddr: Unclosed [...] in pattern');

      // Extract the [groupName] from in-between the [ and ] ...
      const groupName = template.slice(start, i);
      // ... and form a regex from it!
      rx += `(?<${groupName}>[^/]+)`;
    } else {
      rx += escape(template[i]);
    }
  }

  // End of regex
  rx += '$';
  return new RegExp(rx);
};

export const formAddrObject = (matches: RegExpMatchArray, input: Object): Object => {
  const output = input;

  for (const [key, value] of Object.entries(matches.groups!)) {
    output[key] = value
  }

  return output;
};



/**
 * Replaces every bracketed token `[key]` in a template path with the
 * corresponding property from `addr`.
 *
 * @param template A path template containing `[key]` placeholders.
 * @param addr     An object whose keys supply the placeholder values.
 *
 * @example
 * formPath(
 *   "snowflake/databases/[database]/[schema]/external_tables/[name].sql",
 *   { database: "tea_types", schema: "public", name: "oolong" }
 * );
 * -> "snowflake/databases/tea_types/public/external_tables/oolong.sql"
 */
export function formPath(
  template: string,
  addr: Object
): string {
  return template.replace(/\[([^\]]+)]/g, (_, key: string) => {
    if (!(key in addr)) {
      throw new Error(`Missing value for placeholder [${key}]`);
    }
    return String(addr[key]);
  });
}

File: autoschematic-sdk-python/autoschematic_sdk/addr.py
from __future__ import annotations

from dataclasses import dataclass
import re


def match_addr(template: str) -> re.Pattern[str]:
    """Turn a template containing ``[groupName]`` placeholders into a compiled
    regex whose named groups can be read via ``m.group("groupName")``.

    Each ``[placeholder]`` becomes ``(?P<placeholder>[^/]+)``.
    Static parts are regex-escaped automatically.

    Example::

        rx = match_addr("snowflake/warehouses/[name].sql")
        m  = rx.match("snowflake/warehouses/WH1.sql")
        assert m and m.group("name") == "WH1"
    """
    pattern = "^"
    i = 0
    while i < len(template):
        if template[i] == "[":
            i += 1
            start = i
            while i < len(template) and template[i] != "]":
                i += 1
            if i == len(template):
                raise ValueError("match_addr: Unclosed [...] in pattern")
            group_name = template[start:i]
            pattern += f"(?P<{group_name}>[^/]+)"
            i += 1
        else:
            pattern += re.escape(template[i])
            i += 1
    pattern += "$"
    return re.compile(pattern)


def form_addr_object(match: re.Match[str], input: dict[str, str]) -> dict[str, str]:
    """Merge named groups from a regex match into *input* and return it."""
    output = dict(input)
    groups = match.groupdict()
    output.update(groups)
    return output


def form_path(template: str, addr: dict[str, str]) -> str:
    """Replace every ``[key]`` token in *template* with the value from *addr*.

    Example::

        form_path(
            "snowflake/databases/[database]/[schema]/external_tables/[name].sql",
            {"database": "tea_types", "schema": "public", "name": "oolong"},
        )
        # -> "snowflake/databases/tea_types/public/external_tables/oolong.sql"
    """
    def _replace(m: re.Match[str]) -> str:
        key = m.group(1)
        if key not in addr:
            raise KeyError(f"Missing value for placeholder [{key}]")
        return str(addr[key])

    return re.sub(r"\[([^\]]+)\]", _replace, template)

@dataclass
class PointerToVirtual:
    addr: str

@dataclass
class OutputMap:
    map: dict